File: 400/440 - K-th Smallest in Lexicographical Order.py
class Solution:
    def findKthNumber(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: int
        """
        ret = 1
        k -= 1
        while k:
            curr, inc = ret, ret + 1
            diff = 0
            while curr <= n:
                end = min(inc, n + 1)
                diff += max(end - curr, 0)
                curr *= 10
                inc *= 10
            if diff <= k:
                ret += 1
                k -= diff
            else:
                ret *= 10
                k -= 1
        return ret

File: csharp/440.k-th-smallest-in-lexicographical-order.cs
/*
 * @lc app=leetcode id=440 lang=csharp
 *
 * [440] K-th Smallest in Lexicographical Order
 */
public class Solution {
    public int FindKthNumber(int n, int k) {
        int ret = 1;
        --k;
        while (k > 0) {
            long left = ret, right = ret + 1, diff = 0;
            while (left <= n) {
                diff += Math.Min((long) n + 1, right) - left;
                left *= 10;
                right *= 10;
            }
            if (diff <= k) {
                k -= (int) diff;
                ++ret;
            } else {
                --k;
                ret *= 10;
            }
        }
        return ret;
    }
}

